src/impl_ops.rs

```diff
@@ -152,70 +152,56 @@ impl<A, S, D, B> $trt<B> for ArrayBase<S, D>
 #[doc=$doc]
 /// between the reference `self` and the scalar `x`,
 /// and return the result as a new `Array`.
-impl<'a, A, S, D, B> $trt<B> for &'a ArrayBase<S, D>
-    where A: Clone + $trt<B, Output=A>,
+impl<'a, A, S, D, B, C> $trt<B> for &'a ArrayBase<S, D>
+    where A: Clone + $trt<B, Output=C>,
           S: Data<Elem=A>,
           D: Dimension,
           B: ScalarOperand,
 {
-    type Output = Array<A, D>;
-    fn $mth(self, x: B) -> Array<A, D> {
-        self.to_owned().$mth(x)
+    type Output = Array<C, D>;
+    fn $mth(self, x: B) -> Self::Output {
+        self.map(move |elt| elt.clone() $operator x.clone())
     }
 }
     );
 );
 
-// Pick the expression $a for commutative and $b for ordered binop
-macro_rules! if_commutative {
-    (Commute { $a:expr } or { $b:expr }) => {
-        $a
-    };
-    (Ordered { $a:expr } or { $b:expr }) => {
-        $b
-    };
-}
-
 macro_rules! impl_scalar_lhs_op {
-    // $commutative flag. Reuse the self + scalar impl if we can.
-    // We can do this safely since these are the primitive numeric types
-    ($scalar:ty, $commutative:ident, $operator:tt, $trt:ident, $mth:ident, $doc:expr) => (
-// these have no doc -- they are not visible in rustdoc
-// Perform elementwise
-// between the scalar `self` and array `rhs`,
-// and return the result (based on `self`).
-impl<S, D> $trt<ArrayBase<S, D>> for $scalar
-    where S: DataOwned<Elem=$scalar> + DataMut,
-          D: Dimension,
+    ($scalar:ty, $operator:tt, $trt:ident, $mth:ident, $doc:expr) => (
+/// Perform elementwise
+#[doc=$doc]
+/// between the scalar `self` and array `rhs`,
+/// and return the result (based on `self`).
+impl<A, S, D> $trt<ArrayBase<S, D>> for $scalar
+where
+    $scalar: Clone + $trt<A, Output=A>,
+    A: Clone,
+    S: DataOwned<Elem=A> + DataMut,
+    D: Dimension,
 {
     type Output = ArrayBase<S, D>;
-    fn $mth(self, rhs: ArrayBase<S, D>) -> ArrayBase<S, D> {
-        if_commutative!($commutative {
-            rhs.$mth(self)
-        } or {{
-            let mut rhs = rhs;
-            rhs.unordered_foreach_mut(move |elt| {
-                *elt = self $operator *elt;
-            });
-            rhs
-        }})
+    fn $mth(self, mut rhs: ArrayBase<S, D>) -> ArrayBase<S, D> {
+        rhs.unordered_foreach_mut(move |elt| {
+            *elt = self.clone() $operator elt.clone();
+        });
+        rhs
     }
 }
 
-// Perform elementwise
-// between the scalar `self` and array `rhs`,
-// and return the result as a new `Array`.
-impl<'a, S, D> $trt<&'a ArrayBase<S, D>> for $scalar
-    where S: Data<Elem=$scalar>,
-          D: Dimension,
+/// Perform elementwise
+#[doc=$doc]
+/// between the scalar `self` and array `rhs`,
+/// and return the result as a new `Array`.
+impl<'a, A, S, D, B> $trt<&'a ArrayBase<S, D>> for $scalar
+where
+      $scalar: Clone + $trt<A, Output=B>,
+      A: Clone,
+      S: Data<Elem=A>,
+      D: Dimension,
```

Comment:
This impl somehow now breaks Rust -- see the failed tests -- and causes a recursion errror - for an expression that has type `f32` + `f32` which is quite strange/scary(!)

```
   --> tests/oper.rs:159:48
    |
159 |         .fold(f32::zero(), |acc, (&x, &y)| acc + x * y)
    |                                                ^
    |
    = help: consider adding a `#![recursion_limit="256"]` attribute to your crate (`oper`)
    = note: required because of the requirements on the impl of `Add<&ndarray::ArrayBase<_, _>>` for `f32`
    = note: required because of the requirements on the impl of `Add<&ndarray::ArrayBase<_, _>>` for `f32`
    = note: required because of the requirements on the impl of `Add<&ndarray::ArrayBase<_, _>>` for `f32`
    = note: required because of the requirements on the impl of `Add<&ndarray::ArrayBase<_, _>>` for `f32`
    = note: required because of the requirements on the impl of `Add<&ndarray::ArrayBase<_, _>>` for `f32`
    = note: required because of the requirements on the impl of `Add<&ndarray::ArrayBase<_, _>>` for `f32`
    = note: required because of the requirements on the impl of `Add<&ndarray::ArrayBase<_, _>>` for `f32`
    = note: required because of the requirements on the impl of `Add<&ndarray::ArrayBase<_, _>>` for `f32`
```



Unsure if this is a Rust bug, but I think this impl is too general and has infinite descent.

Given the question if `f32` implements `Add<&ArrayBase<S, D>>` look for other impl that has `f32: Add<A>`  where `S: Data<Elem=A>` which looks recursive, is that it?

Reply:
It looks like a compiler bug to me. As you point out, the expression involves only `f32`, but for some reason, the error message indicates that one of the arguments is an array. It's also interesting that on my machine with Rust 1.48.0, the error message is slightly different, saying "impl of `Add<ndarray::ArrayBase<_, _>>` for `f32`" instead of the error message in your comment "impl of `Add<&ndarray::ArrayBase<_, _>>` for `f32`". (Note the `&`.)

The function fails to compile (with the same error message) even after adding type annotations:

```rust
fn reference_dot<'a, V1, V2>(a: V1, b: V2) -> f32
where
    V1: AsArray<'a, f32>,
    V2: AsArray<'a, f32>,
{
    let a: ArrayView1<'a, f32> = a.into();
    let b: ArrayView1<'a, f32> = b.into();
    a.iter()
        .zip(b.iter())
        .fold(f32::zero(), |acc: f32, (&x, &y): (&f32, &f32)| acc + x * y)
}
```

but if I remove the `+ x * y`, it compiles successfully:

```rust
fn reference_dot<'a, V1, V2>(a: V1, b: V2) -> f32
where
    V1: AsArray<'a, f32>,
    V2: AsArray<'a, f32>,
{
    let a: ArrayView1<'a, f32> = a.into();
    let b: ArrayView1<'a, f32> = b.into();
    a.iter()
        .zip(b.iter())
        .fold(f32::zero(), |acc: f32, (&x, &y): (&f32, &f32)| acc)
}
```

I don't see any reason other than a compiler bug for the first function to fail to compile when the second one compiles without errors, since the type annotations confirm that the closure is operating only on `f32` values.

This also compiles successfully:

```rust
fn reference_dot2<'a>(a: ArrayView1<'a, f32>, b: ArrayView1<'a, f32>) -> f32 {
    a.iter()
        .zip(b.iter())
        .fold(f32::zero(), |acc: f32, (&x, &y): (&f32, &f32)| acc + x * y)
}
```

so the bug involves the `.into()` calls in some way. It's surprising that adding explicit type annotations for the results of the `.into()` calls, as in the first example, doesn't work around the bug.

Fwiw, I don't think `impl<'a, A, S, D, B> $trt<&'a ArrayBase<S, D>> for $scalar` is infinitely recursive, since AFAIK it's not possible to have an array of (arrays of (arrays of (arrays of ... [infinite depth]))). The innermost array type can only have an element type that's not an array. You're right that there is recursion if you're dealing with arrays of arrays, but that's the correct behavior, and the recursion is not infinite.

For the particular function we're looking at, the impl doesn't apply, and I don't think the compiler should be trying to apply it. (I think it should only apply the impl if it knows the RHS has some type `&ArrayBase<?S, ?D>`, where `?S` and `?D` are inference variables.)

Reply:
Interesting, the test runners for cross_test, stable, mips vs i686 disagree about the error too, in the same way, even if they both use Rust 1.48

Reply:
I reported the issue (with a simplified example) at rust-lang/rust#80542.